### PR TITLE
[FSDP] Add workaround to fix `buffer_dtype` without root parameters

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143989

Fixes https://github.com/pytorch/pytorch/issues/143900

cc @H-Huang @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/fsdp/_runtime_utils.py [hunk 1]
```diff
@@ -534,7 +534,13 @@ def _root_pre_forward(
         if handle:
             should_cast_buffers_to_full_prec = handle._force_full_precision
         else:
-            should_cast_buffers_to_full_prec = True
+            # If the root has no handle (no managed parameters), then we fall
+            # back to checking if any child wants to force full precision as a
+            # workaround
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:539]: This is a hack, but the code is somewhat beyond saving.

The root FSDP module is responsible for this cast logic, but each FSDP submodule could have different values for `_force_full_precision`, in which case the result is wrong. We could hard error here in that case.

**Review by H-Huang — APPROVED** [review]
Nice quick fix!

**Pull request closed without merging**